burnman/equilibrate.py

```diff
@@ -0,0 +1,775 @@
+    for (code, constraint) in equality_constraints:
+        if code == 'X':
+            c_vector = parameter_constraints_to_endmember_constraint_vector(constraint, prm.indices)
+            A = A.col_join(c_vector.T)
+            b = np.append(b, np.array([0.]))
+            vals = sorted([(i, v) for i, v in enumerate(reduced_nullspace * c_vector)], key=lambda x: np.abs(x[1]))
+            row_indices, values = zip(*vals)
+            a = Matrix(values[:-1])/values[-1]
+            reduced_nullspace = reduced_nullspace[[row_indices[:-1]]] - np.outer(a, reduced_nullspace[[row_indices[-1]]])
+
+    # Convert stoichiometric matrix from endmember to the reduced site formalism using the raw constraint matrix R.
+    # This allows us to use nonnegative least squares.
+
+    # x' = R.x, x' > 0
+    # therefore Rinv.x' = x
+    # We want to solve A.x = b
+    # This is equivalent to A.Rinv.x' = b, for which nnls is appropriate
+    R = simplify_matrix(prm.raw_constraint_matrix)
+    Rinv = (R.H * R) ** -1 * R.H # the left inverse of R
+
+    A = A*Rinv
+    baseline_proportions = nnls(A, b)
+    eps = 1.e-12
+    if  baseline_proportions[1] > eps :
+        raise Exception( "Composition cannot be represented by the given minerals.")
+    # Now we can move within the nullspace (still subject to the non-negativity
+    # constraints) to minimize the difference between the preferred and guessed
+    # solution compositions
+    # A(x + N^T.v) = b
+    # A(Rinv.x' + N^T.v) = b
+    # ARinv(x' + R.N^T.v) = b
+
+    baseline_proportions = np.abs(baseline_proportions[0])
+    if len(reduced_nullspace) != 0:
+        reaction_vectors = reduced_nullspace.T
+        num = np.eye((prm.raw_constraint_matrix.shape[1]))
+        denom = np.eye((prm.raw_constraint_matrix.shape[1]))
+        guess_proportions = np.ones((denom.shape[1]))
+        i=0
+        for phase_idx, mbr_indices in enumerate(prm.indices):
+            n_mbrs = len(mbr_indices)
+            denom[i:i+n_mbrs] = np.einsum('ij->j', denom[i:i+n_mbrs])
+            if n_mbrs > 1:
+                guess = assemblage.phases[phase_idx].guess[mbr_indices]
+                guess_proportions[i:i+n_mbrs] = guess/np.sum(guess)
+            i+=n_mbrs
+        equality_matrix = num - np.einsum('ij, i->ij', denom, guess_proportions)
+        
+        # convert back to simple endmembers
+        baseline_endmember_amounts = Rinv.dot(baseline_proportions)
+
+        cons_fun = lambda x:  R.dot(baseline_endmember_amounts + reaction_vectors.dot(x)) # >= 0
+        cons = ({'type': 'ineq', 'fun': cons_fun})
+
+        sol = minimize(func_min_cdist, [0.]*reaction_vectors.shape[1],
+                       args=(guess_proportions, baseline_endmember_amounts,
+                             reaction_vectors, prm.indices),
+                       method='COBYLA', constraints=cons, tol=1.e-16)
+
+        #assert(sol.status==1)
+        assert(cons_fun(sol.x) > -eps)
```

Comment:
python3 doesn't like comparing a list to a float. Is it enough to check that the minimum of the cons_fun output is > -eps?